sccp/figures/figureS2.py

```diff
@@ -0,0 +1,150 @@
+"""
+S2: Initial Attempt at Pf2 on the lupus data
+article: https://www.science.org/doi/10.1126/science.abf1970
+data: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE174188
+"""
+
+# GOAL: test Pf2 on lupus data, get visualizations for factor matrices
+
+# load functions/modules ----
+from .common import (
+    subplotLabel,
+    getSetup,
+    reorder_table
+)
+from ..parafac2 import parafac2_nd, Pf2X
+from ..imports.scRNA import tensorFy
+import pandas as pd
+import numpy as np
+import seaborn as sns
+import anndata
+from .common import subplotLabel, getSetup
+import matplotlib.pyplot as plt
+import scipy.cluster.hierarchy as sch
+import tensorly as tl
+
+
+####################################################################################
+## LOAD IN DATA -- CAN BE MOVED TO imports/scRNA 
+####################################################################################
+
+def load_lupus_data(third_axis= "ind_cov", n_rand = 100):
+    """Import Lupus PBMC dataset."""
+    X = anndata.read_h5ad("/opt/andrew/lupus/lupus.h5ad")
+    
+    # reorder X so that all of the patients are in alphanumeric order. this is important
+    # so that we can steal cell typings at this point
+    obsV = X.obs_vector('ind_cov')
+    sgUnique, sgIndex = np.unique(obsV, return_inverse=True)
+
+    ann_data_objects = [X[sgIndex == sgi, :] for sgi in range(len(sgUnique))]
+
+    X = anndata.concat(ann_data_objects, axis = 0)
+
+    # select n_rand random patients; evenly split between lupus and SLE
+    #@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+    # REMOVE THIS SECTION TO HAVE ALL INDIVIDUALS INCLUDED (and n_rand arg in function def)
```

Comment:
Doesn't it make more sense to keep all patients, but downsample the cells? I thought I had that in the code I passed along. That way, any downstream analysis isn't affected.

I suppose maybe you did this because of that one patient with fewer cells. Another way to downsample the dataset would be to remove cells only from the patients with the most. e.g., set an upper bound that patients with more than N cells have their data cut to N cells.

Reply:
Yeah downsampling cells ran into the issue where we were using a number of cells that wasn't much bigger than the number of cell states. If we can do every 4th cell (instead of every 10th) this would probably be fine, I can also definitely try downsampling from those with the most cells. I think our biggest concern when selecting cells, rather than patients, was that cell types that are less common but present in all patients might not be included or might be included with an unfair weight? But with this many observations that might not make a difference -- @andrewram4287  let me know if you have other thoughts on that too.

Here's a histogram of counts per patient if that helps to get an idea of what's going on:



![image](https://github.com/meyer-lab/scCP/assets/126696611/458a6827-3bd0-4f54-9332-67f8b527dada)

Reply:
Why don't you do every fourth cell for now, since that should be simple to implement, and then we can remove that when I address the scaling? Even every third cell might fit into memory just fine.

You're right to be concerned about downsampling in any form. Indeed—you could lose rare populations. However, it's probably worse to subsample the patients, because you only have a couple hundred of those. Just like more cells is an opportunity to see rare populations, each patient is an opportunity for the model to build (or refute) its evidence for a pattern being associated with disease.